### PR TITLE
Implement tracking for the number of blueprint backups

### PR DESCRIPTION
**Pull request by luuquangvu**
## Summary by Sourcery

Track and expose the number of blueprint backups, and improve backup restore validation and tooling.

New Features:
- Track the number of existing backups per blueprint and expose it via coordinator metadata and update entity state attributes.
- Provide synchronous and asynchronous helpers to count and check the existence of blueprint backup files, including a service-level check before restore operations.

Bug Fixes:
- Return richer, structured error details for failed blueprint restore operations, including invalid version, missing backup, security violations, and filesystem errors, so callers receive accurate translation placeholders.

Enhancements:
- Refine backup rotation and restore logic to use centralized backup path construction and cached backup counting, and keep in-memory backup counts in sync after save and restore.
- Update the restore blueprint admin service to delegate validation to the coordinator and to surface translation kwargs directly in ServiceValidationError.
- Improve the compatibility validation helper to install Home Assistant with an overrides file, avoiding dependency resolution issues and cleaning up temporary files afterward.

CI:
- Adjust the CI workflow to install Home Assistant under test with a uv overrides file to control dependency resolution.

Tests:
- Extend coordinator, entity, and UI tests to cover backup counting, backup existence checks, invalid version error details, and restore handler behavior across multiple coordinators.
- Add a dedicated backup counting and restore validation test module to validate edge cases such as non-contiguous backups, max_bak limits, and missing backups.

## Summary by CodeRabbit

* **New Features**
  * Blueprint entities now surface a backups_count; backup counting, rotation and existence checks are improved and bounded.

* **Bug Fixes**
  * Restore operations validate requested backup versions and return clear, parameterized localized errors for missing or invalid backups and system failures; translation keys/placeholders standardized.

* **Tests**
  * Expanded coverage for backup counting, existence checks, restore validation, and entity attribute exposure.

* **Chores**
  * CI/install tooling updated to use per-run dependency overrides.

[Review Change Stack](https://app.coderabbit.ai/change-stack/luuquangvu/blueprints-updater/pull/80?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by luuquangvu** [comment]
@sourcery-ai review

File: custom_components/blueprints_updater/coordinator.py [hunk 1]
```diff
@@ -3822,17 +3869,14 @@ def scan_blueprints(
...
 
     @staticmethod
-    def _execute_restore_file(real_path: str, version: int, max_backups: int) -> tuple[bool, str]:
+    def _execute_restore_file(
+        real_path: str, version: int, max_backups: int
+    ) -> tuple[bool, str, int]:
         """Atomic filesystem operation for restoration (runs in executor)."""
         bak_path = f"{real_path}.bak.{version}"
-        if version == 1 and not os.path.isfile(bak_path):
-            legacy_name = f".{os.path.basename(real_path)}.bak"
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/coordinator.py:3829]: The legacy backup method was released early but quickly replaced by a new one, so now it's time to drop it.

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

File: custom_components/blueprints_updater/coordinator.py [hunk 2]
```diff
@@ -3822,17 +3885,11 @@ def scan_blueprints(
...
-        bak_path = f"{real_path}.bak.{version}"
-        if version == 1 and not os.path.isfile(bak_path):
-            legacy_name = f".{os.path.basename(real_path)}.bak"
-            legacy_bak = os.path.join(os.path.dirname(real_path), legacy_name)
-            if os.path.isfile(legacy_bak):
-                bak_path = legacy_bak
-
-        if not os.path.isfile(bak_path):
-            return False, "missing_backup"
+        bak_path = BlueprintUpdateCoordinator._get_backup_path(real_path, version)
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/coordinator.py:3922]: The legacy backup method was released early but quickly replaced by a new one, so now it's time to drop it.

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

File: custom_components/blueprints_updater/coordinator.py [hunk 3]
```diff
@@ -1318,6 +1325,40 @@ async def async_fetch_blueprint(self, path: str, force: bool = False) -> None:
...
+    @staticmethod
+    def _get_backup_path(file_path: str, version: int | str) -> str:
+        """Construct the path to a specific backup file version."""
+        return f"{file_path}.bak.{version}"
+
+    @staticmethod
+    @cache
+    def _count_backups_sync(file_path: str, max_bak: int) -> int:
+        """Count the number of existing backup files for a given blueprint path."""
+        count = 0
```
Comment (SourceryAI) [inline custom_components/blueprints_updater/coordinator.py]: **issue (bug_risk):** The order of `@staticmethod` and `@cache` decorators will prevent `cache_clear()` from working and may break caching.

With `@staticmethod` above `@cache`, `BlueprintUpdateCoordinator._count_backups_sync` returns the `staticmethod` wrapper, while `cache_clear` lives on the inner cached function. Calling `BlueprintUpdateCoordinator._count_backups_sync.cache_clear()` will therefore raise `AttributeError`.

Please either:
- Swap the decorators so `@cache` is outermost, or
- Remove `@staticmethod` and keep a plain/module-level cached function.

This ensures `cache_clear()` is callable on the actual cached function and keeps backup counts correct.

✅ Addressed in 7cb955be3c93012d309bd72fda4c5fedf1cf8d99: Caching is implemented on the module-level `_count_backups_sync_helper` with `@lru_cache`, and `cache_clear()` is called on that helper directly, so there is no longer a decorator-order issue on a `@staticmethod` and cache invalidation works correctly.

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 1 issue, and left some high level feedback:

- In `_execute_restore_file` the legacy backup name handling (`.<basename>.bak` for version 1) has been removed; if older backups may still exist on disk you might want to preserve that fallback to avoid breaking restores of pre-existing backups.
- The restore flow now mixes several hard-coded error strings (e.g. in `async_restore_blueprint` and `async_restore_blueprint_handler`) into `translation_kwargs`; consider centralizing these error keys/strings so they stay consistent with the translation keys and are easier to maintain.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In `_execute_restore_file` the legacy backup name handling (`.<basename>.bak` for version 1) has been removed; if older backups may still exist on disk you might want to preserve that fallback to avoid breaking restores of pre-existing backups.
- The restore flow now mixes several hard-coded error strings (e.g. in `async_restore_blueprint` and `async_restore_blueprint_handler`) into `translation_kwargs`; consider centralizing these error keys/strings so they stay consistent with the translation keys and are easier to maintain.

## Individual Comments

### Comment 1
<location path="custom_components/blueprints_updater/coordinator.py" line_range="1328-1337" />
<code_context>
         )
         self.async_set_updated_data(self.data)

+    @staticmethod
+    def _get_backup_path(file_path: str, version: int | str) -> str:
+        """Construct the path to a specific backup file version."""
</code_context>
<issue_to_address>
**issue (bug_risk):** The order of `@staticmethod` and `@cache` decorators will prevent `cache_clear()` from working and may break caching.

With `@staticmethod` above `@cache`, `BlueprintUpdateCoordinator._count_backups_sync` returns the `staticmethod` wrapper, while `cache_clear` lives on the inner cached function. Calling `BlueprintUpdateCoordinator._count_backups_sync.cache_clear()` will therefore raise `AttributeError`.

Please either:
- Swap the decorators so `@cache` is outermost, or
- Remove `@staticmethod` and keep a plain/module-level cached function.

This ensures `cache_clear()` is callable on the actual cached function and keeps backup counts correct.
</issue_to_address>
~~~

</details>

***

Hi @luuquangvu! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

File: custom_components/blueprints_updater/coordinator.py [hunk 4]
```diff
@@ -3822,17 +3912,11 @@ def scan_blueprints(
...
-            legacy_bak = os.path.join(os.path.dirname(real_path), legacy_name)
-            if os.path.isfile(legacy_bak):
-                bak_path = legacy_bak
-
-        if not os.path.isfile(bak_path):
-            return False, "missing_backup"
+        bak_path = BlueprintUpdateCoordinator._get_backup_path(real_path, version)
 
         try:
             with open(bak_path, encoding="utf-8") as f:
```
[the remark this answers is not in the document]
Comment (luuquangvu) [reply custom_components/blueprints_updater/coordinator.py:3925]: As mentioned earlier, legacy backup was only available in the first few versions, back when the user base was almost non-existent, and it was quickly replaced by numbered backup. Now, after many updates, I want to completely eliminate legacy backup.

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review

**Comment by luuquangvu** [comment]
@coderabbitai full review

**Comment by luuquangvu** [comment]
@sourcery-ai review
